### DISCUSSION TITLE
Massive performance issues with huge maps on modern mobile devices caused by Canvas

### DISCUSSION THREAD

**Discussion by liveman1** [openseadragon/openseadragon#2097] [Q&A]
Hi,

I have massive performance issues for huge maps on modern mobile devices (iPad air 4, iPhone 12), not being able to pan & zoom properly. When zooming in to a detailed section performance is great again. So this happens only on the initial zoom level with the entire image loaded.
There are no issues on older devices like iPhone SE or iPhone 11.
Thus, I guess this is related to the higher resolution of modern displays trying to load & handle the image in a much higher resolution.

I build a huge map from DZI files like this:
```
    var viewer = OpenSeadragon({
        id: "genes",
        tileSources: "genes.dzi",
        ...
        });
	
viewer.addTiledImage({
      	tileSource: 'g1.dzi',
	    x: 0.1,
	    y: 0.1,
	    width: 0.01
	});  
viewer.addTiledImage({
      	tileSource: 'g2.dzi',
	    x: 0.2,
	    y: 0.1,
	    width: 0.01
	});  
viewer.addTiledImage({
      	tileSource: 'g3.dzi',
	    x: 0.3,
	    y: 0.1,
	    width: 0.01
	});  
...
viewer.addTiledImage({
      	tileSource: 'g100.dzi',
	    x: 0.9,
	    y: 0.9,
	    width: 0.01
	});  
```

First of all, is this the right way to build a map?

I already experimented with the proposed solutions from #1938 and #541

Setting `OpenSeadragon.pixelDensityRatio = 0.5;` brings back a good performance. However, everything gets blurred as a huge drawback. Moreover, there are some rendering issues on the overlayed TiledImages like small seams at the edges.
On mobile devices with low resolution displays (e.g. iPhone SE) the image is very sharp & performant.

Do you have any idea how to cope with this issue?
Would it make sense to restrict the overall image resolution on the min zoom level?

Thanks a lot!

**Comment by liveman1** [comment]
I am not sure if the display resolution is causing this.
It perfectly works on iPhone 11 Liquid Retina (1792 x 828 pixel at 326 ppi) and iPhone SE 2. Gen Retina HD(1334 x 750 pixel at 326 ppi).
It does not work on iPhone 12 mini Super Retina (2340 x 1080 at 476 ppi) and iPad Air 4 Liquid Retina (2360 x 1640 pixel at 264 ppi).

No idea...  🤔    At least the behavior is the same across all browsers (Safari, Firefox, Edge)

**Reply by liveman1** [reply]
Hi,

I thought about to remove the added tiles in the first stage & load them on the next zoom level. This way the browser would only have to handle a subset of tiled images & might be more performant.

I tried to figure out how to do this. Its a shame, that I do not have the programming skills to do that :(
I ended up with this:

```
viewer.addTiledImage({
      	tileSource: 'g2.dzi',
	    x: 0.2,
	    y: 0.1,
	    width: 0.01,
            opacity: 0
           });  

viewer.addHandler('bookmark-url-change', function(event) {  
  var zoom = viewer.viewport.getZoom();
  if (zoom >= 2) {
        viewer.world.getItemAt().setOpacity(1)
  } else if (zoom < 2) {
        viewer.world.getItemAt().setOpacity(0)
  }
  });
```

I have not found any example out there. It would be great, if you could help me to get it right.
Thanks!

**Reply by iangilman** [reply]
As mentioned in https://stackoverflow.com/questions/70840661/openseadragon-load-remove-additional-tiled-images-when-zooming-in-out/70854785, something like this should work:

```
// Before this presumably you have made a viewer with a single image.

viewer.addTiledImage({
  tileSource: 'g2.dzi',
  x: 0.2,
  y: 0.1,
  width: 0.01,
  opacity: 0
 });  

viewer.addHandler('zoom', function(event) {  
  var topImage = viewer.world.getItemAt(1);
  if (topImage) {
    var zoom = viewer.viewport.getZoom();
    if (zoom >= 2) {
      topImage.setOpacity(1)
    } else if (zoom < 2) {
      topImage.setOpacity(0)
    }
  }
});
```

**Comment by iangilman** [comment]
I would think that ideally you would just do a single DZI rather than making it out of multiple tiled images. I think the single high-res DZI will always be more efficient. Why are you using multiple? 

Here are a couple more possible issues to explore: https://github.com/openseadragon/openseadragon/issues/1753 and https://github.com/openseadragon/openseadragon/issues/381

**Reply by liveman1** [reply]
Thanks!
Canvas seems to cause the performance issues.
When I set it to false (`useCanvas: false,`) everything works really performant on all iOS devices. 
However, when zooming very deep in the image gets lost & I have to reload the page. This happens also on my Windows PC.
When setting `iOSDevice: 1,` everything works just awesome! 

Is there any explanation for that? Obviously iOSDevice mitigates some iOS issues, however the issue to lose the image happens on Windows PCs as well & gets fixed by setting this property.

Thanks a lot!

**Reply by iangilman** [reply]
Interesting! It looks like `iOSDevice` only affects this one bit of code: 

https://github.com/openseadragon/openseadragon/blob/51e3ea7fa08f10765be25f4c1f9ba7764ee6e79b/src/tiledimage.js#L2054-L2067

That has to do with the `smoothTileEdgesMinZoom` option, which might be the source of your performance issues. Try adding `smoothTileEdgesMinZoom: Infinity` to disable it (see the documentation at http://openseadragon.github.io/docs/OpenSeadragon.html#.Options).

**Reply by liveman1** [reply]
Hi, 
unfortunately, this does not solve the performance issue. When setting to infinity the browser even crashes.

Would it make sense to set canvas to false at min zoom level & set canvas to true at a certain zoom level?
How could this be done? Would be great if you could help me.

`viewer.addHandler('zoom', function(event) {  

  var zoomlevel = 4;
  var zoom = viewer.viewport.getZoom();
    if (zoom >= zoomlevel) {
		  viewer.useCanvas: true;
	  }
	else if (zoom < zoomlevel) {
		  viewer.useCanvas: false;
	  }
}); `

Thanks a lot!

**Reply by iangilman** [reply]
That's very strange that the browser crashes! 

Switching to and from canvas while running doesn't seem like a good idea to me... There's a lot of infrastructure that has to get built and torn down. It would basically be the same as destroying the viewer and starting it again, which of course would create a big discontinuity.

**Reply by liveman1** [reply]
I just tested it again checking all kinds of parameters. However, it really crashes on my iPad/iPhone. Older iPads work fine.
I will now have a screenshot of the assembled image as cover image & load the image tiles on a certain zoom level. 
However, it would be better without this workaround ;)
Is there any logging that I could activate in order to get more insights? Without Inspector within the browser its hard to see what's happening on the iPad.

**Reply by iangilman** [reply]
Do you have a Mac? You can attach an inspector to your iOS device with a Mac using Safari.

**Reply by liveman1** [reply]
unfortunately not
